src/librustc/diagnostics.rs

```diff
@@ -1899,6 +1899,49 @@ contain references (with a maximum lifetime of `'a`).
 [1]: https://github.com/rust-lang/rfcs/pull/1156
 "##,
 
+E0400: r##"
+A user-defined dereference was attempted in an invalid context. Erroneous
+code example:
+
+```
+use std::ops::Deref;
+
+struct A;
+#[derive(Debug)]
+struct B;
+
+impl Deref for A {
+    type Target = B;
+    fn deref(&self)->&B { static B_: B = B; &B_ }
+}
+
+const S: &'static B = &A;
+// error: user-defined dereference operators are not allowed in constants
+```
+
+You cannot store directly a dereference operation results in `const` or `static`
```

Comment:
You cannot directly use a dereference operation whilst initializing a constant or a static. To fix this error, restructure your code to avoid this dereference, perhaps moving it inline: